### PR TITLE
Audit schema coverage

### PR DESCRIPTION
**Pull request by leynos**
## Summary by Sourcery

Introduce a domain-level user-state schema audit for roadmap item 3.5.1, including executable audit logic, tests, and documentation, and mark the roadmap task as complete.

New Features:
- Add a user-state schema audit domain module that evaluates schema snapshots for login, users, profile, and interests coverage and derives migration decisions.
- Expose a schema audit service and helper function via the domain API for use with schema snapshot repositories.

Enhancements:
- Extend domain exports to include user-state schema audit types and service for broader reuse.
- Document the canonical user-state schema audit findings and decisions for roadmap item 3.5.1, and record the associated architecture decision in the backend architecture docs.

Documentation:
- Add an execution plan document for roadmap item 3.5.1 describing scope, constraints, risks, and implementation steps.
- Create a user-state schema audit report documenting coverage findings and migration decisions, and reference it from the architecture documentation.
- Update the backend roadmap to mark item 3.5.1 as complete.

Tests:
- Add unit tests for the user-state schema audit report and service covering baseline, missing-table, and canonical-interests scenarios.
- Add BDD-style behavioural tests and feature definitions exercising the schema audit against an embedded PostgreSQL-backed schema snapshot repository.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[backend/tests/user_state_schema_audit_bdd.rs](https://github.com/leynos/wildside/pull/317/files/b023fed985ff8a28f751dacc86689bfcfb679b09#diff-d373c20efa31aa3fc7881f1dc37d7ae02c96069d33ee9437c116b677fa384e2e)

Comment on file

~~~rust
fn users_and_profile_storage_are_reported_as_covered(world: &mut UserStateSchemaAuditWorld) {
    if skip_if_needed(world) {
        return;
    }

    let report = world.report.as_ref().expect("report should be captured");
    assert_eq!(report.users_coverage, EntitySchemaCoverage::Covered);
    assert_eq!(report.profile_coverage, EntitySchemaCoverage::Covered);
    assert_eq!(
        report.profile_storage_migration,
        MigrationDecision::NotRequired
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/70675/delta?repo-id=1181678&review-id=317&biomarker=Code+Duplication&filename=backend%2Ftests%2Fuser_state_schema_audit_bdd.rs&method=)
The module contains 4 functions with similar structure: interests_migration_decisions_are_not_required,interests_migration_decisions_are_required,users_and_profile_migrations_are_required,users_and_profile_storage_are_reported_as_covered

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- There are unresolved merge conflict markers left in `docs/wildside-backend-architecture.md` (e.g., `<<<<<<< ours` / `||||||| original` / `>>>>>>> theirs`) that need to be resolved and cleaned up before merging.
- The login credential coverage check in `has_login_credential_storage` relies on a small hard-coded set of column and table names; consider centralizing these patterns or documenting them clearly so future schema changes don’t silently break the audit heuristics.

## Individual Comments

### Comment 1
<location path="backend/src/domain/user_state_schema_audit_tests.rs" line_range="13" />
<code_context>
+use crate::domain::ports::{MockSchemaSnapshotRepository, SchemaSnapshotRepositoryError};
+
+#[rstest]
+fn baseline_schema_reports_login_gap_and_interests_migrations() {
+    let report = UserStateSchemaAuditReport::evaluate(&baseline_diagram());
+
</code_context>
<issue_to_address>
**suggestion (testing):** Add explicit tests for the `CredentialsPersisted` login coverage paths

This only tests the `MissingCredentialStorage` path. Please also add unit tests for the `CredentialsPersisted` path in `assess_login_coverage`/`has_login_credential_storage`, covering at least:

1. A `users` table with a recognised credential column (e.g. `password_hash`).
2. A separate credential table (e.g. `user_credentials` with `user_id` and `password_hash`) and no credential columns on `users`.

These will validate both detection strategies and guard against regressions if the heuristics change.

Suggested implementation:

```rust
use rstest::rstest;

use super::{
    EntitySchemaCoverage, InterestsStorageCoverage, LoginSchemaCoverage, MigrationDecision,
    UserStateSchemaAuditReport, UserStateSchemaAuditService, audit_user_state_schema_coverage,
};
use crate::domain::er_diagram::{SchemaColumn, SchemaDiagram, SchemaTable};
use crate::domain::ports::{MockSchemaSnapshotRepository, SchemaSnapshotRepositoryError};

#[rstest]

```

```rust
#[rstest]
fn users_table_with_password_hash_is_recognised_as_credentials_persisted() {
    // users table has an inline credential column (password_hash)
    let diagram = SchemaDiagram {
        tables: vec![SchemaTable {
            name: "users".to_string(),
            columns: vec![
                // primary key / identifier for the user
                SchemaColumn {
                    name: "id".to_string(),
                    data_type: "uuid".to_string(),
                    is_nullable: false,
                    is_primary_key: true,
                },
                // recognised credential column
                SchemaColumn {
                    name: "password_hash".to_string(),
                    data_type: "text".to_string(),
                    is_nullable: false,
                    is_primary_key: false,
                },
            ],
        }],
    };

    let report = UserStateSchemaAuditReport::evaluate(&diagram);

    assert_eq!(
        report.login_schema_coverage,
        LoginSchemaCoverage::CredentialsPersisted
    );
}

#[rstest]
fn separate_user_credentials_table_is_recognised_as_credentials_persisted() {
    // users table has no credential columns; credentials live in a separate table
    let diagram = SchemaDiagram {
        tables: vec![
            SchemaTable {
                name: "users".to_string(),
                columns: vec![SchemaColumn {
                    name: "id".to_string(),
                    data_type: "uuid".to_string(),
                    is_nullable: false,
                    is_primary_key: true,
                }],
            },
            SchemaTable {
                name: "user_credentials".to_string(),
                columns: vec![
                    // foreign key back to users.id (heuristic will treat this as a user reference)
                    SchemaColumn {
                        name: "user_id".to_string(),
                        data_type: "uuid".to_string(),
                        is_nullable: false,
                        is_primary_key: false,
                    },
                    // recognised credential column
                    SchemaColumn {
                        name: "password_hash".to_string(),
                        data_type: "text".to_string(),
                        is_nullable: false,
                        is_primary_key: false,
                    },
                ],
            },
        ],
    };

    let report = UserStateSchemaAuditReport::evaluate(&diagram);

    assert_eq!(
        report.login_schema_coverage,
        LoginSchemaCoverage::CredentialsPersisted
    );
}

#[rstest]

```

The exact constructors/fields for `SchemaDiagram`, `SchemaTable`, and `SchemaColumn`, as well as the field access for `report.login_schema_coverage`, may differ in your codebase. To integrate this cleanly you may need to:
1. Adjust the struct initialisation to match the real `er_diagram` types (e.g. use builder methods like `SchemaTable::new` / `SchemaColumn::new` if those exist, or correct field names and types).
2. Update the primary/foreign key flags to whatever your schema model uses (`is_pk`, `nullable`, etc.).
3. If `UserStateSchemaAuditReport` exposes `login_schema_coverage` via an accessor method instead of a public field, update the assertions to call that method (e.g. `report.login_schema_coverage()`).
4. If `evaluate` is not an associated function on `UserStateSchemaAuditReport` in your code, change the invocation to the appropriate entry point that the other tests use (for example, `audit_user_state_schema_coverage(&diagram, ...)`).
These adjustments will ensure the new tests compile while still modelling the two credential-persistence strategies the comment requested.
</issue_to_address>

### Comment 2
<location path="backend/src/domain/user_state_schema_audit_tests.rs" line_range="62" />
<code_context>
+}
+
+#[rstest]
+fn canonical_revisioned_interests_model_needs_no_interests_migrations() {
+    let report = UserStateSchemaAuditReport::evaluate(&canonical_interests_diagram());
+
</code_context>
<issue_to_address>
**suggestion (testing):** Add tests covering all `InterestsStorageCoverage` variants, including edge cases for revision tracking

Right now we only exercise `DualModel` (baseline) and `CanonicalPreferences` with a `revision` column. Please add focused tests that build diagrams for:

- `InterestsStorageCoverage::Missing` (no interests tables),
- `InterestsStorageCoverage::CanonicalJoinTable` with and without `revision`,
- `DualModel` where a `revision` column exists on one or both tables, asserting `assess_interests_revision_tracking` still returns `false`.

This will fully cover `assess_interests_storage_coverage`, `assess_interests_revision_tracking`, and the `decision_for_*` helpers, and clarify behavior for mixed/partial revisions.

Suggested implementation:

```rust
use rstest::rstest;

```

```rust
use crate::domain::ports::{MockSchemaSnapshotRepository, SchemaSnapshotRepositoryError};

#[rstest]
fn interests_storage_coverage_missing_when_no_interests_tables() {
    let diagram = diagram_without_interests_tables();

    let service = UserStateSchemaAuditService::new(&diagram);

    assert_eq!(
        service.assess_interests_storage_coverage(),
        InterestsStorageCoverage::Missing
    );
    assert_eq!(service.assess_interests_revision_tracking(), false);
}

#[rstest]
fn interests_storage_coverage_canonical_join_without_revision() {
    let diagram = canonical_join_interests_diagram(false);

    let service = UserStateSchemaAuditService::new(&diagram);

    assert_eq!(
        service.assess_interests_storage_coverage(),
        InterestsStorageCoverage::CanonicalJoinTable
    );
    assert_eq!(service.assess_interests_revision_tracking(), false);
}

#[rstest]
fn interests_storage_coverage_canonical_join_with_revision() {
    let diagram = canonical_join_interests_diagram(true);

    let service = UserStateSchemaAuditService::new(&diagram);

    assert_eq!(
        service.assess_interests_storage_coverage(),
        InterestsStorageCoverage::CanonicalJoinTable
    );
    assert_eq!(service.assess_interests_revision_tracking(), true);
}

#[rstest]
fn dual_model_interests_with_mixed_revisions_does_not_track_revisions() {
    // revision on user_interests only
    let diagram_user_only = dual_model_interests_diagram(true, false);
    let service_user_only = UserStateSchemaAuditService::new(&diagram_user_only);
    assert_eq!(
        service_user_only.assess_interests_storage_coverage(),
        InterestsStorageCoverage::DualModel
    );
    assert_eq!(service_user_only.assess_interests_revision_tracking(), false);

    // revision on profile_interests only
    let diagram_profile_only = dual_model_interests_diagram(false, true);
    let service_profile_only = UserStateSchemaAuditService::new(&diagram_profile_only);
    assert_eq!(
        service_profile_only.assess_interests_storage_coverage(),
        InterestsStorageCoverage::DualModel
    );
    assert_eq!(service_profile_only.assess_interests_revision_tracking(), false);

    // revision on both tables
    let diagram_both = dual_model_interests_diagram(true, true);
    let service_both = UserStateSchemaAuditService::new(&diagram_both);
    assert_eq!(
        service_both.assess_interests_storage_coverage(),
        InterestsStorageCoverage::DualModel
    );
    assert_eq!(service_both.assess_interests_revision_tracking(), false);
}

```

To make these tests compile and behave as intended, you will also need to:

1. Implement or expose helper constructors for the diagrams:
   - `diagram_without_interests_tables()`: returns a `SchemaDiagram` identical to your existing baseline diagrams but *without* any tables or join tables that represent interests.
   - `canonical_join_interests_diagram(has_revision: bool)`: returns a `SchemaDiagram` with a single canonical join table for interests. When `has_revision` is `true`, the join table should include the `revision` column in the appropriate schema; when `false`, it should omit it.
   - `dual_model_interests_diagram(user_has_revision: bool, profile_has_revision: bool)`: returns a `SchemaDiagram` corresponding to your `InterestsStorageCoverage::DualModel` shape, toggling the presence of a `revision` column on the user-side and profile-side interests tables according to the boolean flags.

2. Ensure `UserStateSchemaAuditService::new(&SchemaDiagram)` exists and that it exposes:
   - `fn assess_interests_storage_coverage(&self) -> InterestsStorageCoverage`
   - `fn assess_interests_revision_tracking(&self) -> bool`
   matching the behavior implied by the tests and your existing implementation.

3. If you already have diagram builders like `canonical_interests_diagram()` or existing baseline diagrams, consider implementing the three helpers above as thin wrappers around them (e.g. cloning and removing/adding interests tables or `revision` columns) to keep the schema construction consistent across tests.

4. If the audit API uses differently named methods (for example, `decision_for_interests_storage()` or `is_interests_revision_tracked()`), adjust the method names in the assertions accordingly while preserving the same coverage intent: verifying all `InterestsStorageCoverage` variants and the revision tracking behavior for each configuration (missing, canonical join with/without revision, and dual model with mixed revision presence).
</issue_to_address>

### Comment 3
<location path="docs/wildside-backend-architecture.md" line_range="717-720" />
<code_context>
 > plus idempotency replay metadata for offline mutations. This keeps
 > idempotency policy and ownership checks in domain services while outbound
-> repositories remain the only layer with SQL/persistence details.
+> repositories remain the only layer with SQL/persistence details.<<<<<<< ours
 >
 > **Design decision (2026-02-24):** Roadmap item 3.4.1 introduces the
</code_context>
<issue_to_address>
**issue (bug_risk):** Unresolved merge conflict markers need to be removed and the intended text chosen.

This block still includes Git conflict markers (`<<<<<<<`, `|||||||`, `=======`, `>>>>>>>`). Resolve the conflict by choosing the correct text and remove all marker lines before merging.
</issue_to_address>

### Comment 4
<location path="docs/execplans/backend-3-5-1-audit-schema-coverage.md" line_range="391" />
<code_context>
+
+- All commands are re-runnable. Log file names are deterministic per branch
+  and action.
+- If bootstrap or tests fail due transient DB state, cleanly stop embedded
+  PostgreSQL, rerun setup, and re-run only the failing target first before
+  full gates.
</code_context>
<issue_to_address>
**nitpick (typo):** Minor grammar fix: use "due to transient DB state".

Use: `If bootstrap or tests fail due to transient DB state, ...` for correct grammar.

```suggestion
- If bootstrap or tests fail due to transient DB state, cleanly stop embedded
```
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
backend/src/domain/user_state_schema_audit_tests.rs: 

### What lead to degradation?
The module contains 4 functions with similar structure: canonical_join_interests_with_revision_need_no_interests_migrations,canonical_join_interests_without_revision_require_revision_migrations,canonical_revisioned_interests_model_needs_no_interests_migrations,missing_interests_tables_require_interests_migrations

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

File: backend/src/domain/user_state_schema_audit.rs [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+use crate::domain::er_diagram::{SchemaDiagram, SchemaTable};
+use crate::domain::ports::{SchemaSnapshotRepository, SchemaSnapshotRepositoryError};
+
+/// Recognised credential columns used by login persistence heuristics.
+///
+/// Keep this list in sync with schema conventions and migration decisions
+/// documented in `docs/user-state-schema-audit-3-5-1.md`.
+const LOGIN_CREDENTIAL_COLUMNS: &[&str] = &["password_hash", "password_digest", "credential_hash"];
+/// Recognised credential tables used by login persistence heuristics.
+const LOGIN_CREDENTIAL_TABLES: &[&str] = &["user_credentials", "login_credentials", "credentials"];
```
[the remark this answers is not in the document]
Comment (leynos) [reply backend/src/domain/user_state_schema_audit.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
